### PR TITLE
Refine mongo planner tests

### LINKED ISSUE #2879: assumeReadType is not applied to `Subset` branches
Consider this query and its Mongo-specific QScript

```sql
select * from zips where length(city) < 4
```

```
QScript (Mongo-specific):
Map
├─ Filter
│  ├─ ShiftedRead(/jps/zips, ExcludeId)
│  ╰─ Guard
│     ├─ ProjectField
│     │  ├─ ○
│     │  ╰─ Constant(Str(city))
│     ├─ Type(Str)
│     ├─ Lt
│     │  ├─ Length
│     │  │  ╰─ ProjectField
│     │  │     ├─ ○
│     │  │     ╰─ Constant(Str(city))
│     │  ╰─ Constant(Int(4))
│     ╰─ Undefined
╰─ ○
```

(the Map nop here is separate problem)

Adding a `LIMIT` (as the REPL does with `summaryCount`) generates more complex QScript:

```sql
select * from zips where length(city) < 4 limit 10
```

```
QScript (Mongo-specific):
Map
├─ Subset(Take)
│  ├─ Unreferenced
│  ├─ From
│  │  ╰─ Filter
│  │     ├─ ShiftedRead(/jps/zips, ExcludeId)
│  │     ╰─ Guard
│  │        ├─ Guard
│  │        │  ├─ ○
│  │        │  ├─ Type(Obj(Map[], Some(Top)))
│  │        │  ├─ ProjectField
│  │        │  │  ├─ ○
│  │        │  │  ╰─ Constant(Str(city))
│  │        │  ╰─ Undefined
│  │        ├─ Type(Str)
│  │        ├─ Guard
│  │        │  ├─ ○
│  │        │  ├─ Type(Obj(Map[], Some(Top)))
│  │        │  ├─ Lt
│  │        │  │  ├─ Length
│  │        │  │  │  ╰─ ProjectField
│  │        │  │  │     ├─ ○
│  │        │  │  │     ╰─ Constant(Str(city))
│  │        │  │  ╰─ Constant(Int(4))
│  │        │  ╰─ Undefined
│  │        ╰─ Undefined
│  ╰─ Count
│     ╰─ Map
│        ├─ Unreferenced
│        ╰─ Constant(Int(10))
╰─ Guard
   ├─ ○
   ├─ Type(Obj(Map[], Some(Top)))
   ├─ ○
   ╰─ Undefined
```

The cause of this problem is that `assumeReadType` is not applied to `Subset`.

### PR DESCRIPTION
**Pull request by jsantos17**
These are some changes I had sitting in my #2541 branch. They fix a few tests. Also fixes #2879 as it was making it hard to debug #2541.

@rintcius this one has a cleaner history :) Also, I grepped around for a test for either `elideMoreGeneralGuards` or `assumeReadType` and couldn't find any.

### CODE REVIEW

File: mongodb/src/test/scala/quasar/physical/mongodb/planner.scala [hunk 1]
```diff
@@ -857,21 +857,18 @@ class PlannerSpec extends
...
-    }.pendingWithActual(notOnPar, testFile("plan filter with LIKE and OR"))
+           Selector.And(
+             Selector.Doc(
+               BsonField.Name("bar") -> Selector.Type(BsonType.Text)),
+             Selector.Or(
+               Selector.Doc(
+                 BsonField.Name("bar") -> Selector.Regex("^A.*$", false, true, false, false)),
+               Selector.Doc(
+                 BsonField.Name("bar") -> Selector.Regex("^Z.*$", false, true, false, false)))))))
+    }
```
Comment (rintcius) [inline mongodb/src/test/scala/quasar/physical/mongodb/planner.scala]: see https://github.com/quasar-analytics/quasar/pull/2867#discussion_r142617277

File: mongodb/src/test/scala/quasar/physical/mongodb/planner.scala [hunk 2]
```diff
@@ -1389,7 +1377,7 @@ class PlannerSpec extends
           $project(
             reshape("name" -> $field("name")),
             ExcludeId)))
-    }.pendingWithActual(notOnPar, testFile("plan simple sort with field not in projections"))
+    }.pendingWithActual("#2771", testFile("plan simple sort with field not in projections"))
```
Comment (rintcius) [inline mongodb/src/test/scala/quasar/physical/mongodb/planner.scala]: also still has a map reduce

File: mongodb/src/test/scala/quasar/physical/mongodb/planner.scala [hunk 3]
```diff
@@ -1242,14 +1230,14 @@ class PlannerSpec extends
           $project(
             reshape("value" -> $field("bar")),
             ExcludeId)))
-    }.pendingWithActual(notOnPar, testFile("plan simple sort with field in projection"))
+    }.pendingWithActual("#2771", testFile("plan simple sort with field in projection"))
```
[the remark this answers is not in the document]
Comment (jsantos17) [reply mongodb/src/test/scala/quasar/physical/mongodb/planner.scala]: Must be something else then. Will leave it as `notOnPar`.

File: mongodb/src/test/scala/quasar/physical/mongodb/planner.scala (hunk 2 above)
Comment (jsantos17) [reply mongodb/src/test/scala/quasar/physical/mongodb/planner.scala]: Looks like the mapReduce is coming from the sort. I'll investigate a little more later and open an issue if necessary.

**Comment by jsantos17** [comment]
@rintcius ready for another look.

File: mongodb/src/test/scala/quasar/physical/mongodb/planner.scala [hunk 4]
```diff
@@ -1242,14 +1230,14 @@ class PlannerSpec extends
...
 
     "plan simple sort with wildcard" in {
       plan(sqlE"select * from zips order by pop") must
         beWorkflow0(chain[Workflow](
           $read(collection("db", "zips")),
           $sort(NonEmptyList(BsonField.Name("pop") -> SortDir.Ascending))))
-    }.pendingWithActual(notOnPar, testFile("plan simple sort with wildcard"))
+    }.pendingWithActual("#2771", testFile("plan simple sort with wildcard"))
```
[the remark this answers is not in the document]
Comment (wemrysi) [reply mongodb/src/test/scala/quasar/physical/mongodb/planner.scala]: I think it is due to the `ProjectField` in the sort key. Only very straightforward field projections are supported in the aggregation pipeline atm. Would be interesting to see what the legacy planner did here, we might be able to do better but I haven't investigated.

[the remark this answers is not in the document]
Comment (jsantos17) [reply mongodb/src/test/scala/quasar/physical/mongodb/planner.scala]: The QScript for this particular query could be simplified quite a bit with normalization:

```
QScript (Prefer Projection):
Map
├─ Sort
│  ├─ Map
│  │  ├─ ShiftedRead(/jps/zips, ExcludeId)
│  │  ╰─ ConcatArrays
│  │     ├─ MakeArray
│  │     │  ╰─ ○
│  │     ╰─ MakeArray
│  │        ╰─ ○
│  ├─ Bucket
│  │  ╰─ List
│  ╰─ Order
│     ╰─ Ascending
│        ╰─ By
│           ╰─ Guard
│              ├─ ProjectIndex
│              │  ├─ ○
│              │  ╰─ Constant(Int(1))
│              ├─ Type(Obj(Map[], Some(Top)))
│              ├─ ProjectField
│              │  ├─ ProjectIndex
│              │  │  ├─ ○
│              │  │  ╰─ Constant(Int(1))
│              │  ╰─ Constant(Str(pop))
│              ╰─ Undefined
╰─ ProjectIndex
   ├─ ○
   ╰─ Constant(Int(0))
```

I'm not sure if `assumeReadType` has been applied to the sort key branch.

[the remark this answers is not in the document]
Comment (jsantos17) [reply mongodb/src/test/scala/quasar/physical/mongodb/planner.scala]: After some superficial investigation I believe `assumeReadType` applied to the sorting key tree can bypass `MapReduce` entirely for this query. The guard for `Type(Obj(Map[], Some(Top)))` is forcing us into JS and such a guard is not necessary for Mongo.

**Pull request merged**